### ISSUE TITLE
How to use this with async?

### ISSUE THREAD

**Issue by togakangaroo** [Suremaker/LightBDD#8]
If I'm testing something that is async, the step has to return a `Task`. There seems to be no overload for handling this?

I'm using Nunit

**Comment by Suremaker** [comment]
Hey, at this moment LightBDD does not support async step syntax.
While it is possible to declare steps with return types (if [extended step syntax](https://github.com/Suremaker/LightBDD/wiki/03-Scenario-Steps-Definition#extended-step-syntax) is used), any returned value is simply ignored, so async methods will not work.

It means that with current implementation, the step executing asynchronous method will have to synchronously wait for it with Task.WaitAll() or similar method, unfortunately.

The reason I have not implemented it so far was that:
* none of the supported testing frameworks were not offering a real asynchronous processing of the tests (at least then I was checking it). It means that while they were offering a facility to run asynchronous tests, they run all the tests synchronously. As an example, the sample test takes 4 seconds to run, while it should take 2 if it would be run asynchronously,
* there was no requirement like that so far.

```c#
[TestFixture]
public class AsyncTest
{
    [Test]
    public async Task TestOne()
    {
        await Task.Delay(2000);
    }

    [Test]
    public async Task TestTwo()
    {
        await Task.Delay(2000);
    }
}
```

If you are finding this feature needed in LightBDD, I can spend some time implementing it, so it would be available in a next LightBDD version.
If you are however expecting to run all the tests asynchronously to speed them up, then I'm afraid you may get no benefits from async steps.

**Comment by togakangaroo** [comment]
I think you  might be confused between parallel processing and async. Async code is usually not parallel, and parallel processing is not always done asynchronously. The await keyword does not do anything in parallel, it merely tells the task scheduler that you intended this call to take some time and that if it wants to optimize things (for example by returning the thread to the thread pool) that is ok. It is closer to the concept of coroutines or promises from other languages than anything parallel.

So the issue comes up if I'm testing something that has an interface that returns a Task. For example

```
public class TimeEntryCreationService {
     public async Task Create(TimeEntryModel model) {
          if(null != await repository.FindAsync(model.Id))
               throw new InvalidOperationException();
          //...other stuff
          //...maybe other async stuff
          //...etc
      }
}
```

The proper way to invoke the above is with `await service.Create(model)`. But that means that the method containing that call must be async. If it returns a `Task` then it cannot be tested with LightBDD, if it returns `void` then, not only does that go against the microsoft guidance, but also makes it impossible for anything calling that method (ie LIghtBDD) to ever know when that method completes so you cannot run steps in sequence.

This is something that NUnit has supported since 2.6 (I think) and xUnit since 2.0

**Comment by Suremaker** [comment]
> I think you might be confused between parallel processing and async. Async code is usually not parallel, and parallel processing is not always done asynchronously. The await keyword does not do anything in parallel, it merely tells the task scheduler that you intended this call to take some time and that if it wants to optimize things (for example by returning the thread to the thread pool) that is ok. It is closer to the concept of coroutines or promises from other languages than anything parallel.

I am aware of the differences between async and parallel processing. Neither await nor Task.Delay() have anything with parallel. The thing I wanted to highlight there was that Nunit is processing tests synchronously (so it waits for first test to finish, before it starts another one).
So to illustrate, it is executing tests more like this way:
```C#
...
await TestOne();
await TestTwo();
...
```
not like:
```C#
await Task.WhenAll(TestOne(), TestTwo()); //this allows both methods to execute asynchronously
```

Anyway, it is a digression from the main topic...

> The proper way to invoke the above is with await service.Create(model). But that means that the method containing that call must be async. If it returns a Task then it cannot be tested with LightBDD, if it returns void then, not only does that go against the microsoft guidance, but also makes it impossible for anything calling that method (ie LIghtBDD) to ever know when that method completes so you cannot run steps in sequence.

As I mentioned before, it is not supported currently by LightBDD. I'll do some research how to include it in new version.

The only workaround to write steps calling an asynchronous methods in current version of LightBDD would be something like that:
```C#
void When_time_entry_is_created()
{
    _timeEntryCreationService.Create(model).Wait();
}
```
what I know is not the same as using ``await`` but I cannot find any better solution until LightBDD supports async steps properly.

**Comment by togakangaroo** [comment]
Thanks for the clarification, you are talking about parallel *test execution*. As far as I know that used to (4 years ago) be the default for MsTest and is supported [to some degree in xUnit](https://xunit.github.io/docs/running-tests-in-parallel.html). However, my experience with parallel tests is almost entirely negative. Every time I've used that I end up having issues with file locks, or static singletons, etc.

I think the solution is just to use the async stuff built into NUnit and not worry about it from there. This for example is the default in xUnit and their bdd framework (xBehave).

I'll take  a look today too if I have time, it really should just be a fairly simple matter.

**Comment by togakangaroo** [comment]
After investigation I can confirm that this will be either impossible or difficult to do with .Net Framework 4. While we can use the await keyword (as that's a c# thing), you cannot actually await a Task as it didn't have a `GetAwaiter` method until .Net 4.5 and there's no way to polyfil this as the `INotifyCompletion` interface does not exist in .Net 4 either.

What this means is that `LightBDD.NUnit.Tests` will have to be updated to .Net >=4.5 and that possibly the entire stack will have to be updated. I'm going to try to figure out a way to get it to work without updating everything but it might be difficult.

**Comment by togakangaroo** [comment]
Spent a few hours playing around with this. It's not working fully yet but [this is roughly what you would get](https://github.com/surgeforward/LightBDD/tree/adding-async). Obviously it is not refactored and cleaned up yet, but i'm out of time to play with this today.

Note that it is not strictly necessary to update anything but the tests themselves to a recent version of .Net because you can do all the same stuff as `await` using the `Task.ContinueWith` method. **However** I would strongly recommend updating as `await` does several things and I'm sure i got some edge conditions wrong. It also requires some really awkward code (eg recursion just to enumerate a loop). If you could `await` it would also be much easier to get these things right and to refactor cleanly.

At the moment I've added ability to `RunAsync` directly to the `AbstractBDDRunner` but I don't think this is strictly necessary. If for example you wanted to expose this ability for NUnit only, you could put it in the NUnit `BDDRunner` so long as you made the `NewScenario` method protected rather than private.

**Comment by Suremaker** [comment]
Thanks for the investigation and sample code.
I have not had a chance to play with it by myself, but I had done some reading that theoretically it should be possible to enable async/await support for .NET 4.0 by including Microsoft.Bcl.Async packages (http://stackoverflow.com/questions/9110472/using-async-await-on-net-4, https://blogs.msdn.microsoft.com/bclteam/2012/10/22/using-asyncawait-without-net-framework-4-5/) - I have not checked it yet, but I'm going to try it tonight yet.

> At the moment I've added ability to RunAsync directly to the AbstractBDDRunner but I don't think this is strictly necessary.

I think that AbstractBDDRunner is a proper place for it, because it is a general behaviour and it would be good to give users ability to use it, no matter which testing framework is used.

If it would be possible to integrate this behaviour nicely, I would even think to refactor the synchronous methods to use this as well (to eliminate need of maintaining two execution engines) - but first lets see how the implementation goes...

Finally, there is one thing we'll need to be careful about:
* there is a feature of [commenting steps](https://github.com/Suremaker/LightBDD/wiki/09-Step-Comments) that internally uses ExecutionContext class which has [ThreadStatic] field. This will not work with async execution model, so I'll have to find a new way of binding the execution context with scenarios,
* all methods have to be thread safe, because LightBDD supports parallel scenario execution if test framework supports it (XUnit, MbUnit and soon NUnit 3), which means that the ExecutionContext has to support both, async and parallel processing...

I do not know yet how this could be done properly - my first idea is to do something with CallContext (but I see an [article](http://blog.stephencleary.com/2013/04/implicit-async-context-asynclocal.html) discouraging using it with .net 4.0). The second idea is to use SynchronizationContext, but I am not sure if this is a good approach as well...

**Comment by Suremaker** [comment]
Ok, I have uploaded my progress to [async branch](https://github.com/Suremaker/LightBDD/tree/async).

I managed to use async/await keywords with .NET 4.0. All the projects are however dependent on Microsoft.Bcl.Async package...

I still have a few unsolved issues:
* I tried to implement ExecutionContext with SynchronizationContext but it does not work (it looks that [this](http://stackoverflow.com/questions/16109936/synchronizationcontext-is-not-flowed-when-using-await) is the problem :( ),
* It looks like stack trace for exceptions is partially lost (I guess it is related to how I implemented ScenarioBuilder.Run() methods - I need to investigate it),
* I have not checked how it behaves with parameterized scenarios.

I'll try to spend some time tomorrow after work on that...

**Comment by Suremaker** [comment]
Regarding stack trace, it may be this one: http://stackoverflow.com/questions/17921659/await-with-net-4-0-meaningful-stack-traces

**Comment by Suremaker** [comment]
I pushed a second approach to implement the async tasks.
With it, I reverted the Microsoft.Bcl.Async packages and followed your approach of ContinueWith() task chaining.
With this commit, it looks like exception stack trace is a bit better. The SynchronizationContext works also to some extent.
The downside of this implementation is that now, even the synchronous tests are being executed in other threads (StepsConverter uses Task.Factory.StartNew() to call the steps within the task context). 

I will need to spend a few more days on that and rethink which approach would be valid.
I also started wondering that it may just not work well with .net 4 (there is a well known SynchronizationContext propagation issue, code for ContinueWith() is not very readable and I had to implement a workaround to fix exception stack trace) - it looks that it may be better to drop .net 4 support, but then I'll need to switch major version of the project, for what I'd need more time.

**Comment by Suremaker** [comment]
The async support has been implemented in LightBDD 2.0:
- packages information: https://github.com/Suremaker/LightBDD#installation
- documentation: https://github.com/Suremaker/LightBDD/wiki/Scenario-Steps-Definition#asynchronous-scenario-step-execution
